specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1366 @@
+    /// the package stays resolved for that user until the last reference across
+    /// all processes for that user is removed via PackageDependencyContext.Remove()
+    /// (or process termination).
+    ///
+    /// Calls to Add() can be balanced by a PackageDependencyContext.Remove()
+    /// to remove the entry from the package graph.
+    PackageDependencyContext Add();
+
+    /// Resolve a previously pinned PackageDependency to a specific package and
+    /// add it to the calling process' package graph. Once the dependency has
+    /// been added other code-loading methods (LoadLibrary, CoCreateInstance, etc)
+    /// can find the binaries in the resolved package.
+    ///
+    /// Package resoution is specific to a user. The same package dependency can
+    /// resolve to different packages for different users on a system.
+    ///
+    /// This adds the resulting package to the process' package graph.
+    /// A process' package graph is used to search for DLLs (per Dynamic-Link Library Search Order),
+    /// WinRT objects and other resources; the caller can now load DLLs, activate
+    /// WinRT objects and use other resources from the framework package until
+    /// PackageDependencyContext.Remove() is called (or the process ends).
+    /// The package dependency Id must match a package dependency defined
+    /// for the calling user or the system (via PinForSystem) or an exception is raised.
+    ///
+    /// Each successful call adds the resolve packaged to the
+    /// calling process' package graph, even if already present. There is no
+    /// duplicate 'detection' or 'filtering' applied by the API (multiple
+    /// references to a package is not harmful). Once resolution is complete
+    /// the package stays resolved for that user until the last reference across
+    /// all processes for that user is removed via PackageDependencyContext.Remove()
+    /// (or process termination).
+    ///
+    /// This adds the resolved package to the caller's package graph, per rank.
+    /// A process' package graph is a list of packages sorted by rank in ascending
+    /// order (-infinity...0...+infinity). If package(s) are present in the
+    /// package graph with the same rankrank the resolved package is
+    /// (by default) added after others of the same rank. To add a package
+    /// before others of the same rank, specify PackageDependency.PrependIfRankCollision.
+    ///
+    /// Calls to Add() can be balanced by a PackageDependencyContext.Remove()
+    /// to remove the entry from the package graph.
+    PackageDependencyContext Add(AddPackageDependencyOptions options);
+}
+
+/// A unique identifier for a resolved package dependency
+struct PackageDependencyContextId
+{
+    UInt64 Id;
+};
+
+///TBD
+runtimeclass PackageDependencyContext : ICloseable
+{
+    /// Create an intstance of the package dependency context identified by context
+    PackageDependencyContext(PackageDependencyContextId);
```

Comment:
Constructor parameter name should be `contextId`

Reply:
Although maybe this type should have a `GetFromContextId` instead of a ctor? What's the lifecycle of the `Id` passed in here? Does the newly created object own it now?

Reply:
Lifecycle is until the object is destroyed. There's no 'release' method or other way to transfer ownership or prevent the object from closing the context at object destruction. I could go either way. Is there a common pattern for release/detach?

`ctor(id)` and `static object GetFromId(id)` are equivalent except the latter could fail and return null whereas the former can't fail (would throw an exception). IOW Get vs TryGet semantics. Is one routinely preferred over the other?

In this case functionally there's only Get semantics so ctor() vs GetFromId() are equivalent. What's the common practice? What common practice do we want to embrace in Project Reunion?

Reply:
Rolling with the current design. Will see if API Review has feedback